internal/merkle/tree.go

```diff
@@ -0,0 +1,178 @@
+	}
+	if len(leaves) > 0 {
+		tree.nodes[height] = leaves
+	}
+	tree.calculateRootHash()
+	return tree, nil
+}
+
+// Push adds a new leaf node to the Merkle tree.
+//
+// It places the new leaf in the first available position from left to right
+// to maintain the tree's completeness.
+//
+// If the tree is already full it returns an error.
+func (t *Tree) Push(leaf Hash) error {
+	leafCount := len(t.nodes[t.height])
+	if t.height == 0 || uint64(leafCount) == maxLeaves(t.height) {
+		return errors.New("merkle: reached maximum capacity")
+	}
+
+	t.nodes[t.height] = append(t.nodes[t.height], leaf)
+	t.calculateRootHash()
+	return nil
+}
+
+// PushData adds a new leaf node to the Merkle tree.
+//
+// It creates the leaf by hashing the received data with keccak256 and then
+// places it in the first available position from left to right to maintain
+// the tree's completeness.
+//
+// If the tree is already full it returns an error.
+func (t *Tree) PushData(data []byte) error {
+	leaf := crypto.Keccak256Hash(data)
+	return t.Push(leaf)
+}
+
+// RootHash returns the root hash of the Merkle tree.
+//
+// The root hash is a cryptographic hash representing the entire tree.
+// It is calculated by hashing together the hashes of the nodes at each level,
+// starting from the leaves.
+func (t *Tree) RootHash() Hash {
+	return t.nodes[0][0]
+}
+
+// calculateRootHash recalculates the root hash of the tree.
+//
+// This method updates the hash values at each level of the tree
+// starting from the leaves up to the root. It uses the pristine tree nodes
+// to complete any levels that are not yet full.
+func (t *Tree) calculateRootHash() {
+	// for each level, starting from the leaves
+	for currentLevel := t.height; currentLevel > 0; currentLevel-- {
+		nodeCount := len(t.nodes[currentLevel])
+		nextLevelCount := (nodeCount + 1) / 2
+		nextLevel := make([]Hash, 0, nextLevelCount)
+		// for each pair of nodes, starting from the left
+		for i := 0; i < nodeCount; i += 2 {
+			leftNode := t.nodes[currentLevel][i]
+			rightNode := t.node(currentLevel, uint(i+1))
+			// add parent node to the next level
+			nextLevel = append(
+				nextLevel,
+				crypto.Keccak256Hash(leftNode[:], rightNode[:]),
+			)
+		}
+		t.nodes[currentLevel-1] = nextLevel
+	}
+}
+
+// SiblingsOfLeaf returns the sibling nodes of a leaf in bottom-up order.
+//
+// The siblings are the nodes that are paired with the specified leaf
+// at each level of the tree. These nodes are required to compute
+// the Merkle proof for the leaf.
+//
+// If the index is out of bounds, the method returns an error.
+func (t *Tree) SiblingsOfLeaf(index uint) ([]Hash, error) {
+	if uint64(index) >= maxLeaves(t.height) {
+		return nil, errors.New("merkle: index out of bounds")
+	}
+	siblings := make([]Hash, 0, t.height)
+	for level := t.height; level > 0; level-- {
+		siblingIndex := index - 1
+		if index%2 == 0 {
+			siblingIndex = index + 1
+		}
+		siblings = append(siblings, t.node(level, siblingIndex))
```

Comment:
Shouldn't you be dividing index by 2 at each iteration here?

Reply:
And if so, how is it that this is passing the tests? Shouldn't the tests have caught this?